File: app/src/main/java/com/mkulesh/onpc/iscp/messages/HdmiCecMsg.java
/*
 * Copyright (C) 2018. Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

package com.mkulesh.onpc.iscp.messages;

import android.support.annotation.StringRes;

import com.mkulesh.onpc.R;
import com.mkulesh.onpc.iscp.EISCPMessage;
import com.mkulesh.onpc.iscp.ISCPMessage;

/*
 * HDMI CEC settings
 */
public class HdmiCecMsg extends ISCPMessage
{
    public final static String CODE = "CEC";

    public enum Status implements StringParameterIf
    {
        NONE("N/A", R.string.device_two_way_switch_none),
        OFF("00", R.string.device_two_way_switch_off),
        ON("01", R.string.device_two_way_switch_on),
        TOGGLE("UP", R.string.hdmi_cec_toggle);

        final String code;
        final int descriptionId;

        Status(final String code, final int descriptionId)
        {
            this.code = code;
            this.descriptionId = descriptionId;
        }

        public String getCode()
        {
            return code;
        }

        @StringRes
        public int getDescriptionId()
        {
            return descriptionId;
        }
    }

    private final Status status;

    HdmiCecMsg(EISCPMessage raw) throws Exception
    {
        super(raw);
        status = (Status) searchParameter(data, Status.values(), Status.NONE);
    }

    public HdmiCecMsg(Status level)
    {
        super(0, null);
        this.status = level;
    }

    public Status getStatus()
    {
        return status;
    }

    @Override
    public String toString()
    {
        return CODE + "[" + status.toString() + "]";
    }

    @Override
    public EISCPMessage getCmdMsg()
    {
        return new EISCPMessage('1', CODE, status.getCode());
    }

    @Override
    public boolean hasImpactOnMediaList()
    {
        return false;
    }
}

File: lib/iscp/messages/HdmiCecMsg.dart
/*
 * Enhanced Music Controller
 * Copyright (C) 2019-2023 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

import "../../constants/Strings.dart";
import "../ConnectionIf.dart";
import "../EISCPMessage.dart";
import "../ISCPMessage.dart";
import "EnumParameterMsg.dart";

enum HdmiCec
{
    NONE,
    OFF,
    ON,
    TOGGLE
}

/*
 * HDMI CEC settings
 */
class HdmiCecMsg extends EnumParameterMsg<HdmiCec>
{
    static const String CODE = "CEC";

    static const ExtEnum<HdmiCec> ValueEnum = ExtEnum<HdmiCec>([
        EnumItem.code(HdmiCec.NONE, "N/A",
            descrList: Strings.l_device_two_way_switch_none, defValue: true),
        EnumItem.code(HdmiCec.OFF, "00", dcpCode: "OFF",
            descrList: Strings.l_device_two_way_switch_off),
        EnumItem.code(HdmiCec.ON, "01", dcpCode: "ON",
            descrList: Strings.l_device_two_way_switch_on),
        EnumItem.code(HdmiCec.TOGGLE, "UP",
            descrList: Strings.l_device_two_way_switch_toggle)
    ]);

    HdmiCecMsg(EISCPMessage raw) : super(CODE, raw, ValueEnum);

    HdmiCecMsg.toggle(EnumItem<HdmiCec> s, ProtoType proto) :
            super.output(CODE, _toggle(s.key, proto), ValueEnum);

    HdmiCecMsg._dcp(HdmiCec v) : super.output(CODE, v, ValueEnum);

    @override
    bool hasImpactOnMediaList()
    {
        return false;
    }

    static HdmiCec _toggle(HdmiCec s, ProtoType proto)
    => (proto == ProtoType.ISCP) ? HdmiCec.TOGGLE :
        ((s == HdmiCec.OFF) ? HdmiCec.ON : HdmiCec.OFF);

    /*
     * Denon control protocol
     */
    static const String _DCP_COMMAND = "SSHOS";
    static const String _DCP_COMMAND_EXT = "CON";

    static List<String> getAcceptedDcpCodes()
    => [ _DCP_COMMAND ];

    static HdmiCecMsg? processDcpMessage(String dcpMsg)
    {
        final EnumItem<HdmiCec>? s = ValueEnum.valueByDcpCommand(_DCP_COMMAND + _DCP_COMMAND_EXT, dcpMsg);
        return s != null ? HdmiCecMsg._dcp(s.key) : null;
    }

    @override
    String? buildDcpMsg(bool isQuery)
    => _DCP_COMMAND + (isQuery ? (" " + ISCPMessage.DCP_MSG_REQ) :
        _DCP_COMMAND_EXT + " " + getValue.getDcpCode);
}
